net/Socket.hpp

```diff
@@ -1294,7 +1294,20 @@ class StreamSocket : public Socket,
 
         if (!EnableExperimental || events & POLLIN)
         {
+            // readIncomingData returns false only if the read len is 0.
             closed = !readIncomingData() || closed;
+            if (EnableExperimental)
+            {
+                const int last_errno = errno;
+                if (last_errno == EPIPE || last_errno == ECONNRESET)
```

Comment:
Hmm - why bother? there is great merit in simplicity and having fewer code-paths to test.
Is there a genuine problem going around the loop again to detect a HUP in the next poll ?
Of course - if we treat 0 length reads as an error - then perhaps it makes sense to use these error values too but - lets not add complexity and expand our test-matrix just for fun.
Is there a good, hard reason to have a duplicate closing code-path here ? are there good numbers that show a win ?
If not - I'd not do it myself.

Reply:
As discussed, we will check HUP only when we get no POLLIN. So this is obsolete and I'll remove it.

However, for the write case it's worth noting that at least on my system I see ECONNRESET, which we don't check for, rather than EPIPE. So checking for either of these after the write seems sensible. What happens when we don't detect disconnection is that we make testing exit conditions much harder and unreliable (because depending on timing you get extra activity--or not). For example, in the third PR in this series we count the number of uploads done, and when we don't detect disconnections or `_stop` (see above), we do extra uploads, and the test fails.

I'd love to have all externally-observable activity tested and get to a point where we don't do arbitrary activity.